### ISSUE TITLE
Darwin not supported

### ISSUE THREAD

**Issue by MKCarter** [dockstring/dockstring#141]
Hi,

Trying to install this on my Mac. The conda setup works fine. But when running the tests:

    python tutorials/simple_example.py

I obtain the following error:

    raise DockstringError(f"System '{system_name}' not yet supported")
    dockstring.errors.DockstringError: System 'Darwin' not yet supported

Is there a Mac compatible version available?

**Comment by MKCarter** [comment]
OK, I have a fix for this. Removing the 

     #def get_vina_filename() -> str:
     #    """Return system-dependent name of AutoDock Vina executable."""
     #    system_name = platform.system()
     #    if system_name == 'Linux':
     #        return 'vina_linux'
     #    if system_name == 'Darwin':
     #        return 'vina'
     #    else:
     #        raise DockstringError(f"System '{system_name}' not yet supported")

From the utils.py script and adding the vina executable to the vina_path worked :)

**Comment by AustinT** [comment]
> OK, I have a fix for this. Removing the
> 
> ```
>  #def get_vina_filename() -> str:
>  #    """Return system-dependent name of AutoDock Vina executable."""
>  #    system_name = platform.system()
>  #    if system_name == 'Linux':
>  #        return 'vina_linux'
>  #    if system_name == 'Darwin':
>  #        return 'vina'
>  #    else:
>  #        raise DockstringError(f"System '{system_name}' not yet supported")
> ```
> 
> From the utils.py script and adding the vina executable to the vina_path worked :)

Sounds good. We never tested on Mac or Windows actually, so let us know whether this monkey-patch works for you. We could potentially add proper support for MacOS, but we would need to check that the outputs are the same as for linux (e.g. the random seed stuff could be different).

**Comment by MKCarter** [comment]
The results for the tutorials/tests appear to be consistent between Mac and Linux. I will let you know if anything unexpected occurs.

**Comment by gncs** [comment]
@MKCarter, thank you for raising this issue. Could you try #142? Thank you.

**Comment by MKCarter** [comment]
This is still throwing the Darwin error for me:

```
Traceback (most recent call last):
     File "tutorials/simple_example.py", line 7, in <module>
    score, _ = target.dock("CC1=C(C(=O)N2CCCCC2=N1)CCN3CCC(CC3)C4=NOC5=C4C=CC(=C5)F")
     File "/Users/michaelcarter/DD_tools/test/dockstring/dockstring/target.py", line 189, in dock
    self._dock_pdbqt(ligand_pdbqt, vina_logfile, vina_outfile, seed=seed, num_cpus=num_cpus)
    File "/Users/michaelcarter/DD_tools/test/dockstring/dockstring/target.py", line 120, in _dock_pdbqt
    get_vina_path(),
    File "/Users/michaelcarter/DD_tools/test/dockstring/dockstring/utils.py", line 94, in get_vina_path
    path = get_bin_dir() / get_vina_filename()
    File "/Users/michaelcarter/DD_tools/test/dockstring/dockstring/utils.py", line 65, in get_vina_filename
    raise DockstringError(f"System '{system_name}' not yet supported")
dockstring.errors.DockstringError: System 'Darwin' not yet supported
```

**Comment by gncs** [comment]
Could you please double-check that you are using the correct code version (cf8f4483dfde090de6472559c6447f7714ef6d07) as the error should be raised in line 67, not in line 65.

**Comment by MKCarter** [comment]
You are correct, my mistake. Once running the updated code this did work (sorry conda issues)
This is the expected output - it should be good to merge these changes :) 
`Docking was successful, score=-11.8. END OF SCRIPT.
`

**Comment by AustinT** [comment]
> You are correct, my mistake. Once running the updated code this did work (sorry conda issues) This is the expected output - it should be good to merge these changes :) `Docking was successful, score=-11.8. END OF SCRIPT. `

Before merging we should run `pytest`. I can run the test on my colleague's Mac potentially.

**Comment by MKCarter** [comment]
```
=============================================================================================================== test session starts ===============================================================================================================
platform darwin -- Python 3.7.10, pytest-7.1.2, pluggy-1.0.0
rootdir: /Users/michaelcarter/DD_tools/test/dockstring-add_mac_support, configfile: pytest.ini
collected 55 items                                                                                                                                                                                                                                

tests/test_benchmarks.py ..F..F...                                                                                                                                                                                                          [ 16%]
tests/test_targets.py .......................F..F.F.....F.F.........                                                                                                                                                                        [100%]

==================================================================================================================== FAILURES =====================================================================================================================
___________________________________________ TestOriginalBenchmarks.test_benchmark_values[promiscuous_PPAR-O=C1C=CC2=CC=C3C=C(Nc4cccc(C5C=c6cc7c(cc6=C5)CC=C7)c4)N=C3C12-expected_obj_parts1--11.59705] ____________________________________________

self = <test_benchmarks.TestOriginalBenchmarks object at 0x7f8288ce3c50>
original_benchmarks = {'F2': BenchmarkObjective(base_functions={'F2': functools.partial(<function safe_dock_function at 0x7f8288ce1830>, tar...ction at 0x7f8288ce1830>, target_name='LCK')}, aggregation_function=<function selective_JAK2_score at 0x7f8288cec050>)}
name = 'promiscuous_PPAR', top_smiles = 'O=C1C=CC2=CC=C3C=C(Nc4cccc(C5C=c6cc7c(cc6=C5)CC=C7)c4)N=C3C12', expected_obj_parts = {'PPARA': -13.3, 'PPARD': -13.8, 'PPARG': -13.3, 'QED': 0.829705}, expected_obj = -11.59705

    @pytest.mark.slow
    @pytest.mark.parametrize("name,top_smiles,expected_obj_parts,expected_obj", [
        ("F2", "CC1=CC2(c3cccc(C(=O)NC4=NC(=O)CC(F)=N4)c3)CC(=C1)C=CC2=O", dict(F2=-12.7, QED=0.85955), -11.2955),
        ("promiscuous_PPAR", "O=C1C=CC2=CC=C3C=C(Nc4cccc(C5C=c6cc7c(cc6=C5)CC=C7)c4)N=C3C12",
         dict(PPARA=-13.3, PPARD=-13.8, PPARG=-13.3, QED=0.829705), -11.59705),
        (
            "selective_JAK2",
            "CC1=NCC(C(N)=O)=NN1c1cc(-c2c(C)cc(C)cc2C)ccc1F",
            dict(JAK2=-12.2, LCK=-8.3, QED=0.9182718),
            -11.182718,
        ),
    ])
    def test_benchmark_values(self, original_benchmarks: dict, name: str, top_smiles: str,
                              expected_obj_parts: Dict[str, float], expected_obj: float):
        """
        Test the values of all the benchmark functions using top-scoring SMILES from the paper.
        """
    
        benchmark_fn = original_benchmarks[name]
    
        # First, check that the score formula seems correct
        # (i.e. if the parts of the score are correct, do they form the correct whole)
        assert close_enough(expected_obj, benchmark_fn.aggregation_function(**expected_obj_parts))
    
        # Second, check end-to end whether the SMILES produces the score expected
        actual_obj, actual_obj_parts = benchmark_fn(top_smiles)
        assert close_enough(actual_obj, expected_obj)  # overall value matches
        # Ensure parts are the same up to a certain tolerance
        assert set(actual_obj_parts.keys()) == set(expected_obj_parts.keys())
        for key in actual_obj_parts.keys():
>           assert close_enough(actual_obj_parts[key], expected_obj_parts[key])
E           assert False
E            +  where False = close_enough(-13.7, -13.8)

tests/test_benchmarks.py:57: AssertionError
_____________________________________________________________ TestOriginalBenchmarks.test_benchmark_nan[promiscuous_PPAR-OC1=C(C(N[C@H]2[C@H](N=C(C3=C(O)C=CC=C3)C)CCCC2)=C)C=CC=C1] ______________________________________________________________

self = <test_benchmarks.TestOriginalBenchmarks object at 0x7f8288cf8410>
original_benchmarks = {'F2': BenchmarkObjective(base_functions={'F2': functools.partial(<function safe_dock_function at 0x7f8288ce1830>, tar...ction at 0x7f8288ce1830>, target_name='LCK')}, aggregation_function=<function selective_JAK2_score at 0x7f8288cec050>)}
name = 'promiscuous_PPAR', nan_smiles = 'OC1=C(C(N[C@H]2[C@H](N=C(C3=C(O)C=CC=C3)C)CCCC2)=C)C=CC=C1'

    @pytest.mark.slow
    @pytest.mark.parametrize(
        "name,nan_smiles",
        [
            ("F2", "S=C1N(C(O)C(=C(C2C(=O)N(C(=S)N(C2=O)C)C)C=3C=CN=CC3)C(=O)N1C)C"),
            ("promiscuous_PPAR", "OC1=C(C(N[C@H]2[C@H](N=C(C3=C(O)C=CC=C3)C)CCCC2)=C)C=CC=C1"),  # not NaN for first one
            ("promiscuous_PPAR", "O=C1C2(CN3C4(N(C2)CC1(C3)C(C)C)C5=C(NC4=O)C=CC=C5)C(C)C"),  # nan for all
            ("selective_JAK2", "S=C1N(C(O)C(=C(C2C(=O)N(C(=S)N(C2=O)C)C)C=3C=CN=CC3)C(=O)N1C)C"),  # nan for JAK2 only
            ("selective_JAK2", "O=C1C2(CN3C4(N(C2)CC1(C3)C(C)C)C5=C(NC4=O)C=CC=C5)C(C)C"),  # nan for LCK only
        ])
    def test_benchmark_nan(self, original_benchmarks: dict, name: str, nan_smiles: str):
        """
        Make sure everything is ok for SMILES with NaN docking scores.
        The NaN smiles are sourced from the dockstring dataset itself.
        """
        benchmark_fn = original_benchmarks[name]
    
        # Test 1: does it run ok?
        value, _ = benchmark_fn(nan_smiles)
    
        # Test 2: is the value actually nan?
>       assert math.isnan(value)
E       assert False
E        +  where False = <built-in function isnan>(-5.767804909171941)
E        +    where <built-in function isnan> = math.isnan

tests/test_benchmarks.py:80: AssertionError
_______________________________________________________________________________________________________________ test_simple_docking _______________________________________________________________________________________________________________

    def test_simple_docking():
        """Simple docking test, set to run in all cases."""
        target = load_target('ABL1')
    
        smiles_1 = 'CCO'
        energy_1, _ = target.dock(smiles_1)
>       assert energy_1 is not None and math.isclose(energy_1, -2.4)
E       assert (-2.6 is not None and False)
E        +  where False = <built-in function isclose>(-2.6, -2.4)
E        +    where <built-in function isclose> = math.isclose

tests/test_targets.py:148: AssertionError
_______________________________________________________________________________________________________ TestDocking_Slow.test_seeds[0--4.6] _______________________________________________________________________________________________________

self = <test_targets.TestDocking_Slow object at 0x7f8288e4b650>, seed = 0, score = -4.6

    @pytest.mark.parametrize('seed, score', [(0, -4.6), (1, -4.6), (2, -4.7)])
    def test_seeds(self, seed: int, score: float):
        target = load_target('CYP3A4')
        energy, aux = target.dock(lysine_smiles, seed=seed)
>       assert energy is not None and math.isclose(energy, score)
E       assert (-4.3 is not None and False)
E        +  where False = <built-in function isclose>(-4.3, -4.6)
E        +    where <built-in function isclose> = math.isclose

tests/test_targets.py:173: AssertionError
_______________________________________________________________________________________________________ TestDocking_Slow.test_seeds[2--4.7] _______________________________________________________________________________________________________

self = <test_targets.TestDocking_Slow object at 0x7f8288e4b950>, seed = 2, score = -4.7

    @pytest.mark.parametrize('seed, score', [(0, -4.6), (1, -4.6), (2, -4.7)])
    def test_seeds(self, seed: int, score: float):
        target = load_target('CYP3A4')
        energy, aux = target.dock(lysine_smiles, seed=seed)
>       assert energy is not None and math.isclose(energy, score)
E       assert (-4.6 is not None and False)
E        +  where False = <built-in function isclose>(-4.6, -4.7)
E        +    where <built-in function isclose> = math.isclose

tests/test_targets.py:173: AssertionError
______________________________________________________________________________________________________ TestDocking_Slow.test_positive_score _______________________________________________________________________________________________________

self = <test_targets.TestDocking_Slow object at 0x7f8288e4e150>

    def test_positive_score(self):
        target = load_target('AR')
        smiles = r'O1/C(=N\CCC=2C=3C(NC2)=CC=CC3)/[C@]4(N(C(=O)C5C=6[C@H]4[C@H]7[C@@H](CC6[C@@H]8[C@H]([C@H]5C)C(=O)N(C8=O)C9=CC=CC=C9)C(=O)N(C7=O)C%10=CC=CC=C%10)C1=O)CC%11=CC=CC=C%11'
        score, aux = target.dock(smiles)
>       assert score is not None and score >= 0.0
E       assert (-6.1 is not None and -6.1 >= 0.0)

tests/test_targets.py:198: AssertionError
____________________________________________________________________________________________________ TestDocking_Slow.test_pdbqt_to_pdb_error _____________________________________________________________________________________________________

self = <test_targets.TestDocking_Slow object at 0x7f8288e4e210>

    def test_pdbqt_to_pdb_error(self):
        target = load_target('CYP3A4')
        score, aux = target.dock('O=C1N(C=2N=C(OC)N=CC2N=C1C=3C=CC=CC3)C4CC4')
        affinities = [-9.0, -8.8, -8.4, -8.3, -8.3, -7.9, -7.7, -7.7, -7.6]
        assert aux['ligand'].GetNumConformers() == 9
        assert len(aux['affinities']) == len(affinities)
>       assert all(math.isclose(a, b) for a, b in zip(aux['affinities'], affinities))
E       assert False
E        +  where False = all(<generator object TestDocking_Slow.test_pdbqt_to_pdb_error.<locals>.<genexpr> at 0x7f8288da3650>)

tests/test_targets.py:213: AssertionError
============================================================================================================= short test summary info =============================================================================================================
FAILED tests/test_benchmarks.py::TestOriginalBenchmarks::test_benchmark_values[promiscuous_PPAR-O=C1C=CC2=CC=C3C=C(Nc4cccc(C5C=c6cc7c(cc6=C5)CC=C7)c4)N=C3C12-expected_obj_parts1--11.59705] - assert False
FAILED tests/test_benchmarks.py::TestOriginalBenchmarks::test_benchmark_nan[promiscuous_PPAR-OC1=C(C(N[C@H]2[C@H](N=C(C3=C(O)C=CC=C3)C)CCCC2)=C)C=CC=C1] - assert False
FAILED tests/test_targets.py::test_simple_docking - assert (-2.6 is not None and False)
FAILED tests/test_targets.py::TestDocking_Slow::test_seeds[0--4.6] - assert (-4.3 is not None and False)
FAILED tests/test_targets.py::TestDocking_Slow::test_seeds[2--4.7] - assert (-4.6 is not None and False)
FAILED tests/test_targets.py::TestDocking_Slow::test_positive_score - assert (-6.1 is not None and -6.1 >= 0.0)
FAILED tests/test_targets.py::TestDocking_Slow::test_pdbqt_to_pdb_error - assert False
==================================================================================================== 7 failed, 48 passed in 198.85s (0:03:18) =====================================================================================================

```

**Comment by AustinT** [comment]
Sorry for the silence @MKCarter. It looks like the issue comes from different handling of random seeds between Mac and Linux, which is not something we know how to fix. Because we want to ensure that the output of `dockstring` matches the dataset we published, we don't want to relax the requirements of these tests either. Therefore I have decide to close this issue and mark it `wontfix` for now. However, if somebody wants to help us fix this in the future we would be very happy to provide whatever support is needed.

**Comment by AustinT** [comment]
Since this issue was closed, I got a Macbook myself. I did a bit more testing and it looks like the docking scores don't match the dataset very well. Here is a plot of randomly chosen docking scores from the dataset and the docking scores for the same molecules computed by my Mac

![image](https://github.com/dockstring/dockstring/assets/22754579/54265abe-faf1-4be6-ac0d-9304663683e8)

The scores only match for ~50% of molecules, and the `r^2` is not super high. Unfortunately I don't see an easy way to fix this. *However*, I also think there is no harm in enabling docking on Mac provided it is clear that the docking scores will not match the dataset. I am making a PR with this now.
